src/collapse/xctrace.rs

```diff
@@ -0,0 +1,455 @@
+    fn collapse<R, W>(&mut self, reader: R, writer: W) -> std::io::Result<()>
+    where
+        R: std::io::BufRead,
+        W: std::io::Write,
+    {
+        let mut reader = Reader::from_reader(reader);
+        self.collapse_inner(&mut reader, writer)
+    }
+
+    fn is_applicable(&mut self, input: &str) -> Option<bool> {
+        let mut input = input.as_bytes();
+        let mut line = String::new();
+        loop {
+            if let Ok(n) = input.read_line(&mut line) {
+                if n == 0 {
+                    break;
+                }
+            } else {
+                return Some(false);
+            }
+
+            let trimmed = line.trim();
+            if !trimmed.is_empty() {
+                return Some(trimmed.contains(r#"<?xml version="1.0"?>"#));
+            }
+            line.clear();
+        }
+        None
+    }
+}
+
+impl Folder {
+    fn collapse_inner<R, W>(&mut self, reader: &mut Reader<R>, writer: W) -> io::Result<()>
+    where
+        R: std::io::BufRead,
+        W: std::io::Write,
+    {
+        let mut context = CollapseContext::default();
+        let mut buf = Vec::new();
+        let nodes = loop {
+            let event = match reader.read_event_into(&mut buf) {
+                Ok(x) => x,
+                Err(e) => return invalid_data_error!("Read xml event failed: {:?}", e),
+            };
+            match event {
+                Event::Start(start) => {
+                    let attributes = start.attributes();
+                    let name = start.name().into_inner();
+                    let new_state = match (context.state_backtrace.top_mut(), name) {
+                        (None, TRACE_QUERY_RESULT) => {
+                            CurrentTag::TraceQueryResult(TraceQueryResultState::default())
+                        }
+                        (Some(CurrentTag::TraceQueryResult(_)), NODE) => {
+                            let node_state = NodeState { rows: Vec::new() };
+                            CurrentTag::Node(node_state)
+                        }
+                        (Some(CurrentTag::Node(_)), ROW) => {
+                            let row_state = RowState { backtrace: None };
+                            CurrentTag::Row(row_state)
+                        }
+                        (Some(CurrentTag::Row(_)), BACKTRACE) => {
+                            let id = attributes_to_backtrace(&attributes)?;
+                            CurrentTag::Backtrace(BacktraceState {
+                                id,
+                                frames: Vec::new(),
+                            })
+                        }
+                        (Some(CurrentTag::Backtrace(_)), FRAME) => {
+                            let (id, name) = attributes_to_frame(&attributes)?;
+                            CurrentTag::Frame(FrameState { id, name })
+                        }
+                        _ => CurrentTag::Other(String::from_utf8_lossy(name).into_owned()),
```

Comment:
Should this ever happen? If it happens, should we warn and `continue` instead (perhaps with some special logic to not err in the `End`)?